node/silkworm/stagedsync/stage_execution.cpp

```diff
@@ -109,52 +112,47 @@ StageResult Execution::forward(db::RWTxn& txn) {
-    auto key{db::block_key(from)};
-    auto data{hashes_table.find(db::to_slice(key), true)};
-    while (data.done) {
-        BlockNum reached_block_num{endian::load_big_u64(static_cast<const uint8_t*>(data.key.data()))};
-        if (reached_block_num != from) {
-            throw std::runtime_error("Bad canonical header sequence: expected " + std::to_string(from) + " got " +
-                                     std::to_string(reached_block_num));
-        }
+    assert(prefetched_blocks_.empty());
 
-        Bytes block_key(8 + kHashLength, '\0');
-        std::memcpy(&block_key[0], data.key.data(), 8);
-        std::memcpy(&block_key[8], data.value.data(), kHashLength);
-
-        Block block{};
-        auto raw_header{db::read_header_raw(*txn, block_key)};
-        if (raw_header.empty()) {
-            throw std::runtime_error("Unable to load block header for block " + std::to_string(from));
-        }
-        ByteView encoded_header{raw_header.data(), raw_header.length()};
-        rlp::success_or_throw(rlp::decode(encoded_header, block.header));
+    const size_t count{std::min(static_cast<size_t>(to - from + 1), kMaxPrefetchedBlocks)};
+    size_t num_read{0};
 
-        if (!db::read_body(*txn, block_key, /*read_senders=*/true, block)) {
-            throw std::runtime_error("Unable to load block body for block " + std::to_string(from));
-        }
-        ret.push(block);
-
-        if (from == to || ret.size() >= max_blocks) {
-            break;
-        }
+    db::Cursor hashes_table(txn, db::table::kCanonicalHashes);
+    auto key{db::block_key(from)};
+    if (hashes_table.seek(db::to_slice(key))) {
+        db::WalkFunc walk_function{[&](mdbx::cursor&, mdbx::cursor::move_result& data) {
+            BlockNum reached_block_num{endian::load_big_u64(static_cast<const uint8_t*>(data.key.data()))};
+            if (reached_block_num != from) {
+                throw std::runtime_error("Bad canonical header sequence: expected " + std::to_string(from) + " got " +
+                                         std::to_string(reached_block_num));
+            }
+            SILKWORM_ASSERT(data.value.length() == kHashLength);
+            const auto hash_ptr{static_cast<const uint8_t*>(data.value.data())};
+            prefetched_blocks_.push_back();
+            if (!db::read_block(*txn, gsl::span<const uint8_t, kHashLength>{hash_ptr, kHashLength}, from,
+                                /*read_senders=*/true, prefetched_blocks_.back())) {
+                throw std::runtime_error("Unable to read block " + std::to_string(from));
+            }
+            ++from;
+            return true;
+        }};
+        num_read = db::cursor_for_count(hashes_table, walk_function, count);
+    }
 
-        ++from;
-        data = hashes_table.to_next(false);
+    if (num_read != count) {
+        throw std::runtime_error("Missing block " + std::to_string(from + num_read));
```

Comment:
This shows incorrect info: `from` is altered within walk function so the printed value is not coherent.

Reply:
Indeed. Fixed in https://github.com/torquem-ch/silkworm/commit/18d9e63ce2200e43143ef959c0223e3b33fa2df9